### PR TITLE
Automated generation of streets and district file from Open Street Maps for CEA

### PR DESCRIPTION
**Pull request by JIMENOFONSECA**
This solves issue #1849 and #1851  which introduces three new tools for CEA

To test:

Run 
datamanagement/district-helper.py
datamanagemnet/streets-helper.py
datamanagemnet/terrain-helper.py

It should create respectively the district.shp, the streets.shp and the terrain.tiff for your scenario>

Then run radiation, so to check that the geometry works.

### CODE REVIEW

File: cea/default.config [hunk 1]
```diff
@@ -52,6 +52,42 @@ archetypes.type = MultiChoiceParameter
...
+streets.choices = all_private, walk, bike, drive, drive_service, all
+streets.help = type of streets to gather. Default All_private
+
+[terrain-helper]
+grid-size = 30
+grid-size.type = IntegerParameter
+grid-size.nullable = false
+grid-size.help = Default: 30 m. grid size for raster. this affects the accuracy of reflections from the solar irradiation.
+
+elevation = 0
```
Comment (daren-thomas) [inline cea/default.config]: This sets the default to 0 (as opposed to empty as described in the help text) which one should it be?

File: cea/default.config [hunk 2]
```diff
@@ -52,6 +52,42 @@ archetypes.type = MultiChoiceParameter
...
+streets = all_private
+streets.type = ChoiceParameter
+streets.choices = all_private, walk, bike, drive, drive_service, all
+streets.help = type of streets to gather. Default All_private
+
+[terrain-helper]
+grid-size = 30
+grid-size.type = IntegerParameter
+grid-size.nullable = false
+grid-size.help = Default: 30 m. grid size for raster. this affects the accuracy of reflections from the solar irradiation.
```
Comment (daren-thomas) [inline cea/default.config]: These help texts show up in the dashboard interface. They need to be checked for grammar / spelling.

File: cea/default.config [hunk 3]
```diff
@@ -52,6 +52,42 @@ archetypes.type = MultiChoiceParameter
...
 
+[streets-helper]
+bbox =
+bbox.type = ListParameter
+bbox.help = Default: empty, it uses district shapefile as bounding box, otherwise indicate Bounding box coordinates. bbox = min Longitude , min Latitude , max Longitude , max Latitude, if empty we
+
+streets = all_private
+streets.type = ChoiceParameter
+streets.choices = all_private, walk, bike, drive, drive_service, all
+streets.help = type of streets to gather. Default All_private
```
Comment (daren-thomas) [inline cea/default.config]: These help texts show up in the dashboard interface. They need to be checked for grammar / spelling.

File: cea/default.config [hunk 4]
```diff
@@ -52,6 +52,42 @@ archetypes.type = MultiChoiceParameter
 archetypes.choices = comfort, architecture, HVAC, internal-loads, supply, restrictions
 archetypes.help = List of input databases to create.
 
+[streets-helper]
+bbox =
+bbox.type = ListParameter
+bbox.help = Default: empty, it uses district shapefile as bounding box, otherwise indicate Bounding box coordinates. bbox = min Longitude , min Latitude , max Longitude , max Latitude, if empty we
+
+streets = all_private
+streets.type = ChoiceParameter
```
Comment (daren-thomas) [inline cea/default.config]: I wonder if a MultiChoiceParameter would be more suitabe?

File: cea/default.config [hunk 5]
```diff
@@ -52,6 +52,42 @@ archetypes.type = MultiChoiceParameter
...
+
+[district-helper]
+buffer = 50
+buffer.type = RealParameter
+buffer.help = Default: 50 meters, it is the length of the buffer area surrounding the Zone geometry
+
+height-ag =
+height-ag.type = RealParameter
+height-ag.nullable = true
+height-ag.help = Default: None, it is the average height of buildings. leave empty, we calculate it from Open Street Maps (no accurrate)
```
Comment (daren-thomas) [inline cea/default.config]: These help texts show up in the dashboard interface. They need to be checked for grammar / spelling.

File: cea/datamanagement/district_geometry_helper.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
+"""
```
Comment (daren-thomas) [inline cea/datamanagement/district_geometry_helper.py]: Why don't we call this tool `create-district-geometry`? I think that might communicate the intent a bit better. What do you think? The same for the terrain-helper and the other one...

**Comment by daren-thomas** [comment]
Working with the default config file and test case, the `district-helper` runs through (haven't checked the output) but the `streets-helper` raises an error:

```
(cea) λ cea streets-helper
City Energy Analyst version 2.11
Running `cea streets-helper` with the following parameters:
- general:scenario = C:\reference-case-open\baseline
  (default: C:\reference-case-open\baseline)
- streets-helper:bbox = []
  (default: [])
- streets-helper:streets = all_private
  (default: all_private)
generating streets from District Geometry
Traceback (most recent call last):
  File "c:\Users\darthoma\Miniconda2\envs\cea\Scripts\cea-script.py", line 11, in <module>
    load_entry_point('cityenergyanalyst', 'console_scripts', 'cea')()
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\interfaces\cli\cli.py", line 65, in main
    script_module.main(config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\datamanagement\streets_helper.py", line 117, in main
    geometry_extractor_osm(locator, config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\datamanagement\streets_helper.py", line 69, in geometry_extractor_osm
    network_type=type_of_streets)
  File "c:\Users\darthoma\Miniconda2\envs\cea\lib\site-packages\osmnx\core.py", line 1494, in graph_from_bbox
    G = truncate_graph_bbox(G_buffered, north, south, east, west, retain_all=retain_all, truncate_by_edge=truncate_by_edge)
  File "c:\Users\darthoma\Miniconda2\envs\cea\lib\site-packages\osmnx\core.py", line 1005, in truncate_graph_bbox
    G = get_largest_component(G)
  File "c:\Users\darthoma\Miniconda2\envs\cea\lib\site-packages\osmnx\utils.py", line 336, in get_largest_component
    if not nx.is_weakly_connected(G):
  File "<decorator-gen-198>", line 2, in is_weakly_connected
  File "c:\Users\darthoma\Miniconda2\envs\cea\lib\site-packages\networkx\utils\decorators.py", line 82, in _not_implemented_for
    return not_implement_for_func(*args, **kwargs)
  File "c:\Users\darthoma\Miniconda2\envs\cea\lib\site-packages\networkx\algorithms\components\weakly_connected.py", line 167, in is_weakly_connected
    """Connectivity is undefined for the null graph.""")
networkx.exception.NetworkXPointlessConcept: Connectivity is undefined for the null graph.
```

File: cea/default.config (hunk 1 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: fixed, thanks!

File: cea/default.config (hunk 2 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: fixed, thanks!

File: cea/default.config (hunk 3 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: fixed, thanks!

File: cea/default.config (hunk 4 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: not for the moment. It is supposed to be only one choice for now.

File: cea/default.config (hunk 5 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: fixed, thanks!

File: cea/datamanagement/district_geometry_helper.py (hunk 6 above)
Comment (JIMENOFONSECA) [reply cea/datamanagement/district_geometry_helper.py]: in principle yes, however we have the data-helper already. maybe we change it all later?

**Comment by JIMENOFONSECA** [comment]
@daren-thomas thanks for checking that error. I did the fixes. Please check it again. Thanks!

File: cea/datamanagement/district_geometry_helper.py (hunk 6 above)
Comment (daren-thomas) [reply cea/datamanagement/district_geometry_helper.py]: sure. i like consistency. i couldn't find the issue i once created for the data-helper to be renamed. we can do that someday when we have time...

**Comment by daren-thomas** [comment]
ran the 3 scripts and it seems to work. currently running radiation-daysim but it seems to be working. I think we can merge this.

**Comment by JIMENOFONSECA** [comment]
Do the honors :-))

_______________________

Dr. Jimeno A. Fonseca
Senior researcher and project coordinator
Future cities Laboratory
Singapore ETH-Centre

Chair of Architecure and Building Systems
ETH Zurich

On 4 Apr 2019, 15:54 +0800, Daren Thomas <notifications@github.com>, wrote:

ran the 3 scripts and it seems to work. currently running radiation-daysim but it seems to be working. I think we can merge this.

**Comment by daren-thomas** [comment]
waiting for jenkins ;-)

**Pull request merged**